src/irmin/tree.ml

```diff
@@ -690,53 +693,73 @@ module Make (P : S.PRIVATE) = struct
 
     let empty_marks () = Hashes.create 39
 
-    let fold ~force ~uniq ~pre ~post ~path f t acc =
+    let fold ~force ~uniq ~pre ~post ~path ?depth ~node ~contents t acc =
       let marks =
         match uniq with
         | `False -> dummy_marks
         | `True -> empty_marks ()
         | `Marks n -> n
       in
-      let rec aux ~path acc t k =
-        match force with
-        | `True -> (
-            to_map t >>= function
-            | Ok m -> (map [@tailcall]) ~path acc (Some m) k
-            | Error (`Dangling_hash _) -> (map [@tailcall]) ~path acc None k)
-        | `False skip -> (
-            match t.info.map with
-            | Some n -> (map [@tailcall]) ~path acc (Some n) k
-            | _ -> skip path acc)
-      and aux_uniq ~path acc t k =
-        if uniq = `False then (aux [@tailcall]) ~path acc t k
+      let rec aux ~path acc d t k =
+        let apply acc = node path t acc in
+        let next acc =
+          match force with
+          | `True | `And_clear -> (
+              to_map t >>= function
+              | Ok m ->
+                  if force = `And_clear then clear ~depth:0 t;
+                  (map [@tailcall]) ~path acc d (Some m) k
+              | Error (`Dangling_hash _) -> (map [@tailcall]) ~path acc d None k
+              )
+          | `False skip -> (
+              match t.info.map with
+              | Some n -> (map [@tailcall]) ~path acc d (Some n) k
+              | _ -> skip path acc)
```

Comment:
Hi there :)

Shouldn't `k` be called after skip?

Reply:
Hey! Yes good catch, I'll fix it in another PR :-)

Reply:
See #781